libdrgn: reads from segment mapping

chenyingk opened the pull request:
We added an mmap interface for kcore (https://lore.kernel.org/patchwork/patch/1453867/).
It can reduce system calls significantly when accessing the kernel data structures.
The patch will be merged into next kernel version.

Co-developed-by:Feng Zhou <zhoufeng.zf@bytedance.com>
Signed-off-by: chenying <chenying.kernel@bytedance.com>

osandov commented on libdrgn/memory_reader.c:
```diff
@@ -284,3 +284,33 @@ struct drgn_error *drgn_read_memory_file(void *buf, uint64_t address,
+	char *map_address = file_segment->map_address;
+	size_t file_count;
+
+	if (!map_address)
```
We only use this callback if we successfully mapped this segment, right? So we should either remove this check or get rid of `drgn_read_memory_mmap()` and have `drgn_read_memory_file()` handle both the mmap and read cases.

osandov commented on libdrgn/program.c:
```diff
@@ -16,6 +16,7 @@
 #include <string.h>
 #include <sys/statfs.h>
 #include <unistd.h>
+#include <sys/mman.h>
```
Alphabetical order for this include, please.

```suggestion
#include <string.h>
#include <sys/mman.h>
#include <sys/statfs.h>
#include <unistd.h>
```

osandov commented on libdrgn/program.c:
```diff
@@ -86,6 +87,37 @@ void drgn_program_init(struct drgn_program *prog,
+	uint64_t file_size;
+	void *map_address;
+
+	if (elf_getphdrnum(prog->core, &phnum) != 0)
```
Rather than digging through the core file again, let's just save `num_file_segments` in `struct drgn_program`.

osandov commented on libdrgn/program.c:
```diff
@@ -381,9 +416,18 @@ drgn_program_set_core_dump(struct drgn_program *prog, const char *path)
+
+		map_address = mmap(NULL, phdr->p_filesz, PROT_READ,
+				MAP_PRIVATE, prog->core_fd, phdr->p_offset);
+		if (map_address && map_address != MAP_FAILED) {
```
According to [`mmap(2)`](https://man7.org/linux/man-pages/man2/mmap.2.html), mmap won't return `NULL` unless `MAP_FIXED` is given.

```suggestion
		if (map_address != MAP_FAILED) {
```

osandov reviewed (changes requested):
Sorry for the delay, I was on vacation last week and am catching up.

This is amazing! I wasn't sure that it'd be possible to implement `mmap` for `/proc/kcore` in the kernel.

I have a few small comments on the code, and a couple of more general questions:
1. How well will this work for normal core dump files? My concern is that core dump segments might not be always be aligned to the page size, but at least on the core dumps that I have lying around, they are properly aligned.
2. Should we skip the mmap path on 32-bit systems? Virtual address space is much tighter there, but I suppose that 32-bit systems will usually be debugging 32-bit core dumps that will probably still fit in the address space.

chenyingk commented:
> We added an mmap interface for kcore (https://lore.kernel.org/patchwork/patch/1453867/).
> It can reduce system calls significantly when accessing the kernel data structures.
> The patch will be merged into next kernel version.
> 
> Co-developed-by:Feng Zhou [zhoufeng.zf@bytedance.com](mailto:zhoufeng.zf@bytedance.com)
> Signed-off-by: chenying [chenying.kernel@bytedance.com](mailto:chenying.kernel@bytedance.com)

chenyingk commented:
> Sorry for the delay, I was on vacation last week and am catching up.
> 
> This is amazing! I wasn't sure that it'd be possible to implement `mmap` for `/proc/kcore` in the kernel.
> 
> I have a few small comments on the code, and a couple of more general questions:
> 
> 1. How well will this work for normal core dump files? My concern is that core dump segments might not be always be aligned to the page size, but at least on the core dumps that I have lying around, they are properly aligned.
> 2. Should we skip the mmap path on 32-bit systems? Virtual address space is much tighter there, but I suppose that 32-bit systems will usually be debugging 32-bit core dumps that will probably still fit in the address space.

Thanks for your comments. I will revision addresses your comments in the next few days. But，I'm sorry I don't fully understand your concern about the mmap for 32-bit systems.

chenyingk commented:
All comments have been addressed.
